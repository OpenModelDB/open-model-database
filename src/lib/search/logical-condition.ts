import { assertNever } from '../util';

export type Condition<T> = Condition.Var<T> | Condition.And<T> | Condition.Or<T> | Condition.Not<T> | boolean;
// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace Condition {
    export type Var<T> = { readonly type: 'var'; value: T };
    export type And<T> = { readonly type: 'and'; items: Condition<T>[] };
    export type Or<T> = { readonly type: 'or'; items: Condition<T>[] };
    export type Not<T> = { readonly type: 'not'; item: Condition<T> };

    export function variable<T>(value: T): Condition.Var<T> {
        return { type: 'var', value };
    }
    export function and<T>(...conditions: Condition<T>[]): Condition.And<T> {
        return { type: 'and', items: conditions };
    }
    export function or<T>(...conditions: Condition<T>[]): Condition.Or<T> {
        return { type: 'or', items: conditions };
    }
    export function not<T>(condition: Condition<T>): Condition.Not<T> {
        return { type: 'not', item: condition };
    }

    export function stringify<T>(
        condition: Condition<T> | CompiledCondition<T>,
        toString: (variable: T) => string = String
    ): string {
        if (typeof condition === 'boolean') {
            return condition ? 'T' : 'F';
        }

        switch (condition.type) {
            case 'const':
                return condition.value ? 'T' : 'F';
            case 'var': {
                const value = toString(condition.value);
                if ('negated' in condition && condition.negated) {
                    return `¬${value}`;
                }
                return value;
            }
            case 'not':
                return `¬${stringify(condition.item, toString)}`;
            case 'and':
                return `∀(${condition.items.map((i) => stringify(i, toString)).join(' ')})`;
            case 'or':
                return `∃(${condition.items.map((i) => stringify(i, toString)).join(' ')})`;
            default:
                return assertNever(condition);
        }
    }
}

export type AtLeastTwo<T> = [T, T, ...T[]];
export type CompiledCondition<T> = CompiledVar<T> | CompiledAnd<T> | CompiledOr<T> | CompiledConst;
export type CompiledVar<T> = { readonly type: 'var'; readonly value: T; negated: boolean };
export type CompiledAnd<T> = { type: 'and'; items: AtLeastTwo<CompiledVar<T> | CompiledOr<T>> };
export type CompiledOr<T> = { type: 'or'; items: AtLeastTwo<CompiledVar<T> | CompiledAnd<T>> };
export type CompiledConst = { readonly type: 'const'; value: boolean };

function otherOperation(operation: 'and' | 'or'): 'and' | 'or' {
    return operation === 'and' ? 'or' : 'and';
}
function neutralElementOf(operation: 'and' | 'or'): boolean {
    return operation === 'and';
}
function negateInPlace<T>(condition: CompiledCondition<T>): void {
    if (condition.type === 'var') {
        condition.negated = !condition.negated;
        return;
    }
    if (condition.type === 'const') {
        condition.value = !condition.value;
        return;
    }

    // De-Morgan
    condition.type = otherOperation(condition.type);
    condition.items.forEach(negateInPlace);
}
/**
 * Optimizes the given condition and converts it into a form that is more optimal for evaluation.
 */
export function compileCondition<T>(condition: Readonly<Condition<T>>): CompiledCondition<T> {
    if (typeof condition === 'boolean') {
        return { type: 'const', value: condition };
    }
    if (condition.type === 'var') {
        return { type: 'var', value: condition.value, negated: false };
    }
    if (condition.type === 'not') {
        const c = compileCondition(condition.item);
        negateInPlace(c);
        return c;
    }

    const neutral = neutralElementOf(condition.type);

    // trivial
    if (condition.items.length === 0) {
        return { type: 'const', value: neutral };
    }
    if (condition.items.length === 1) {
        return compileCondition(condition.items[0]);
    }

    const items: (CompiledVar<T> | CompiledAnd<T> | CompiledOr<T>)[] = [];
    const vars = new Set<T>();
    const negVars = new Set<T>();
    for (const item of condition.items.map(compileCondition)) {
        if (item.type === 'const') {
            if (item.value === neutral) {
                // it's neutral, so we don't need to include it
            } else {
                // this means that it's the absorbing element
                return item;
            }
        } else if (item.type === 'var' || item.type === condition.type) {
            for (const i of item.type === 'var' ? [item] : item.items) {
                if (i.type === 'var') {
                    const opposite = i.negated ? vars : negVars;
                    if (opposite.has(i.value)) {
                        // we just found a∧¬a == false or a∨¬a == true
                        return { type: 'const', value: !neutral };
                    }

                    const target = i.negated ? negVars : vars;
                    if (!target.has(i.value)) {
                        // only add non-duplicate variables
                        target.add(i.value);
                        items.push(i);
                    }
                } else {
                    items.push(i);
                }
            }
        } else {
            items.push(item);
        }
    }

    if (items.length === 0) {
        return { type: 'const', value: neutral };
    }
    if (items.length === 1) {
        return items[0];
    }
    return { type: condition.type, items: items as never };
}

export function testCondition<T>(condition: CompiledCondition<T>, resolveVariable: (variable: T) => boolean): boolean {
    const evaluate = (c: CompiledCondition<T>): boolean => {
        switch (c.type) {
            case 'const':
                return c.value;
            case 'var':
                return resolveVariable(c.value) !== c.negated;
            case 'and':
                return c.items.every(evaluate);
            case 'or':
                return c.items.some(evaluate);
            default:
                return assertNever(c);
        }
    };
    return evaluate(condition);
}
