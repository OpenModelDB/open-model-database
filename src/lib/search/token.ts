const numberRe = /^\d+$/;

const knownWords = /(jpeg|png|dds|dxt)/;

export function tokenize(query: string): string[] {
    const tokens = query
        .split(/[\s\p{P}\p{Z}_+.]+|(?=\p{Lu}\p{Ll})|([vV](?=\d)|\d+|[A-Z](?:[A-Z](?![a-z]))+)|\b/u)
        .filter(Boolean)
        .map((s) => s.toLowerCase());

    // split known words
    for (let i = 0; i < tokens.length; i++) {
        const a = tokens[i];

        if (knownWords.test(a)) {
            const t = a.split(knownWords).filter(Boolean);
            tokens.splice(i, 1, ...t);
            i += t.length - 1;
        }
    }

    for (let i = 1; i < tokens.length; i++) {
        const a = tokens[i - 1];
        const b = tokens[i];

        // split of trailing v for versions
        if (a.length > 1 && a.endsWith('v') && numberRe.test(b)) {
            tokens.splice(i - 1, 1, a.slice(0, -1), 'v');
            i--;
            continue;
        }

        // merge certain adjacent tokens
        const join = (a === 'v' && numberRe.test(b)) || (numberRe.test(a) && b === 'x');
        if (join) {
            tokens.splice(i - 1, 2, a + b);
        }
    }

    return tokens;
}

const boundaryRe0 = /^|$|\b|(?=[\s\p{P}\p{Z}_+.])|(?=\p{Lu}\p{Ll})/y;
const boundaryRe1 = /[\s\p{P}\p{Z}_+.]|\p{L}\P{L}|\P{L}\p{L}|\d\D|\D\d|[A-Z][A-Z][a-z]/y;
export function isBoundary(text: string, index: number): boolean {
    boundaryRe0.lastIndex = index;
    if (boundaryRe0.test(text)) return true;

    if (index > 0) {
        // step one char back
        index--;

        // surrogate pair
        const c = text.charCodeAt(index);
        if (c >= 0xdc00 && c <= 0xdfff) {
            index--;
        }

        if (boundaryRe1.test(text)) return true;
    }
    return false;
}
