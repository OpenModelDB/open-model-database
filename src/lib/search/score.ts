import { isBoundary } from './token';

/**
 * Returns a score for how well the given text matches a certain query, the higher the better.
 *
 * A score of 0 or lower indicates the that given text doesn't match the query at all.
 */
export type ScoreFn = (text: string) => number;

export interface ScoringOptions {
    g1BaseScore: number | ((token: string) => number);
    g1StartAtBoundaryBonus: number;
    g1CombineMatchScores: (scores: number[]) => number;
    g1CombineTokenScores: (scores: number[]) => number;
    g2DistExp: number;
}

export const DEFAULT_OPTIONS: Readonly<ScoringOptions> = {
    g1BaseScore: (token) => token.length ** 0.5,
    g1StartAtBoundaryBonus: 4,
    g1CombineMatchScores: (scores) => scores.reduce((a, b) => a + b, 0),
    g1CombineTokenScores: (scores) => scores.reduce((a, b) => a + b, 0),
    g2DistExp: 1,
};

interface G1Match {
    index: number;
    score: number;
}

export function createScoreFn(queryTokens: string[], options: Readonly<ScoringOptions> = DEFAULT_OPTIONS): ScoreFn {
    if (queryTokens.length === 0) {
        // we don't have a query, so we'll just say that every text matches
        return () => 1;
    }

    return (text) => {
        const g1Matches: G1Match[][] = queryTokens.map((token) => {
            const baseScore =
                typeof options.g1BaseScore === 'number' ? options.g1BaseScore : options.g1BaseScore(token);
            return findMatches(text, token).map((index) => {
                const scoreAddition = isBoundary(text, index) ? options.g1StartAtBoundaryBonus : 0;
                return { index, score: baseScore + scoreAddition };
            });
        });

        const g1Score = options.g1CombineTokenScores(
            g1Matches.map((matches) => options.g1CombineMatchScores(matches.map((m) => m.score)))
        );

        return g1Score;
    };
}

function findMatches(text: string, needle: string): number[] {
    if (needle.length === 0) return [];

    const matches: number[] = [];
    let last = 0;
    let i;
    while ((i = text.indexOf(needle, last)) !== -1) {
        matches.push(i);
        last = i + needle.length;
    }

    return matches;
}
/**
 * Returns all pairs of numbers from `a` and `b` such that the average distance
 * `(a[aIndex] - b[bIndex] + c) ** 2` is minimized.
 *
 * Both `a` and `b` as assumed to be sorted and only contain unique values.
 */
function find2GramPairs(a: readonly number[], b: readonly number[], c: number): [aIndex: number, bIndex: number][] {
    if (a.length === 0 || b.length === 0) return [];
    if (b.length < a.length) {
        // we want a to be the shorter array
        return find2GramPairs(b, a, -c).map(([x, y]) => [y, x]);
    }
    if (a.length === 1) {
        return [[0, binarySearchIndexOfClosest(b, a[0] + c)]];
    }

    return a
        .map((v, i) => {
            const closest = binarySearchIndexOfClosest(b, v + c);
            const bValue = b[closest];
            return { index: i, bIndex: closest, dist: Math.abs(v - bValue + c) };
        })
        .sort((a, b) => a.dist - b.dist)
        .map(({ index, bIndex }) => {
            return [index, bIndex];
        });
}

/**
 * Returns the index if the number in `list` closest to `n`.
 */
function binarySearchIndexOfClosest(list: readonly number[], n: number): number {
    if (list.length === 0) {
        return 0;
    }

    let low = 0;
    let high = list.length;
    while (low < high) {
        const m = (low + high) >> 1;
        const v = list[m];
        if (v === n) {
            return m;
        } else if (v < n) {
            low = m + 1;
        } else {
            high = m;
        }
    }

    // the index low might be off by one in either direct, so we have to check
    const i0 = Math.max(0, Math.min(low - 1, list.length - 1));
    const i1 = Math.max(0, Math.min(low, list.length - 1));
    const i2 = Math.max(0, Math.min(low + 1, list.length - 1));

    const d0 = Math.abs(list[i0] - n);
    const d1 = Math.abs(list[i1] - n);
    const d2 = Math.abs(list[i2] - n);

    if (d0 < d1 && d0 < d2) return i0;
    if (d2 < d1 && d2 < d0) return i2;
    return i1;
}
