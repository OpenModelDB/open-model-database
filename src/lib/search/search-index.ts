import { EMPTY_SET, assertNever } from '../util';
import { CompiledCondition, testCondition } from './logical-condition';
import { createScoreFn } from './score';

export interface WeightedText {
    readonly text: string;
    readonly weight: number;
}
export interface CorpusEntry<Id, Tag> {
    readonly id: Id;
    readonly tags: ReadonlySet<Tag>;
    readonly texts: readonly WeightedText[];
}
export interface SearchResult<Id> {
    readonly id: Id;
    /** A score greater than zero. */
    readonly score: number;
}

export class SearchIndex<Id, Tag> {
    readonly entries: ReadonlyMap<Id, CorpusEntry<Id, Tag>>;
    readonly allIds: ReadonlySet<Id>;
    private readonly byTag: ReadonlyMap<Tag, ReadonlySet<Id>>;

    constructor(entries: Iterable<CorpusEntry<Id, Tag>>) {
        const byId = new Map<Id, CorpusEntry<Id, Tag>>();
        const byTag = new Map<Tag, Set<Id>>();

        for (const entry of entries) {
            if (byId.has(entry.id)) {
                throw new Error(`Duplicate id ${String(entry.id)}`);
            }
            byId.set(entry.id, entry);

            for (const tag of entry.tags) {
                let ids = byTag.get(tag);
                if (ids === undefined) {
                    ids = new Set();
                    byTag.set(tag, ids);
                }
                ids.add(entry.id);
            }
        }

        this.entries = byId;
        this.byTag = byTag;
        this.allIds = new Set(byId.keys());
    }

    /**
     * Returns a set of all entries that fulfill the given tag condition.
     *
     * The order of ids in the iterable is not determined and must not be relied upon.
     */
    withTags(condition: CompiledCondition<Tag>): ReadonlySet<Id> {
        switch (condition.type) {
            case 'const': {
                return condition.value ? this.allIds : EMPTY_SET;
            }
            case 'var': {
                const ref = this.byTag.get(condition.value) ?? EMPTY_SET;
                return condition.negated ? without(this.allIds, ref) : ref;
            }
            case 'and': {
                const refs: ReadonlySet<Id>[] = [];
                const negRefs: ReadonlySet<Id>[] = [];
                let complete = true;

                for (const item of condition.items) {
                    if (item.type === 'var') {
                        const ref = this.byTag.get(item.value) ?? EMPTY_SET;
                        (item.negated ? negRefs : refs).push(ref);
                    } else {
                        complete = false;
                    }
                }

                const selection = without(refs.length === 0 ? this.allIds : intersect(...refs), union(...negRefs));
                if (complete) {
                    return selection;
                }
                return setFilter(selection, (id) => {
                    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                    const entry = this.entries.get(id)!;
                    return testCondition(condition, (t) => entry.tags.has(t));
                });
            }
            case 'or': {
                const refs: ReadonlySet<Id>[] = [];
                const negRefs: ReadonlySet<Id>[] = [];
                for (const item of condition.items) {
                    if (item.type === 'var') {
                        const ref = this.byTag.get(item.value) ?? EMPTY_SET;
                        (item.negated ? negRefs : refs).push(ref);
                    } else {
                        refs.push(this.withTags(item));
                    }
                }

                if (negRefs.length > 0) {
                    const neg = intersect(...negRefs);
                    if (neg.size === 0) {
                        return this.allIds;
                    }
                    refs.push(without(this.allIds, neg));
                }
                return union(...refs);
            }
            default:
                return assertNever(condition);
        }
    }

    retrieve(condition: CompiledCondition<Tag>, queryTokens: string[]): SearchResult<Id>[] {
        const selection = this.withTags(condition);
        const scoreFn = createScoreFn(queryTokens);

        const results: SearchResult<Id>[] = [];
        for (const id of selection) {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            const entry = this.entries.get(id)!;
            let score = 0;
            for (const { text, weight } of entry.texts) {
                score += scoreFn(text) * weight;
            }
            if (score > 0) {
                results.push({ id, score });
            }
        }
        return results;
    }
}

function intersect<T>(...sets: ReadonlySet<T>[]): ReadonlySet<T> {
    if (sets.length === 0) {
        throw new Error('Cannot intersect 0 sets');
    }
    if (sets.length === 1) {
        return sets[0];
    }

    sets.sort((a, b) => a.size - b.size);
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const smallest = sets.shift()!;
    if (smallest.size === 0) {
        return EMPTY_SET;
    }

    const result = new Set<T>();

    if (sets.length === 1) {
        const [a] = sets;
        for (const i of smallest) {
            if (a.has(i)) {
                result.add(i);
            }
        }
    } else if (sets.length === 2) {
        const [a, b] = sets;
        for (const i of smallest) {
            if (a.has(i) && b.has(i)) {
                result.add(i);
            }
        }
    } else if (sets.length === 3) {
        const [a, b, c] = sets;
        for (const i of smallest) {
            if (a.has(i) && b.has(i) && c.has(i)) {
                result.add(i);
            }
        }
    } else {
        for (const i of smallest) {
            if (sets.every((s) => s.has(i))) {
                result.add(i);
            }
        }
    }

    return result;
}

function union<T>(...sets: ReadonlySet<T>[]): ReadonlySet<T> {
    // remove all empty sets
    sets = sets.filter((s) => s.size > 0);

    if (sets.length === 0) {
        return EMPTY_SET;
    }
    if (sets.length === 1) {
        return sets[0];
    }

    sets.sort((a, b) => b.size - a.size);
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const largest = sets.shift()!;
    const result = new Set<T>(largest);

    for (const set of sets) {
        for (const i of set) {
            result.add(i);
        }
    }

    return result;
}

function without<T>(a: ReadonlySet<T>, b: ReadonlySet<T>): ReadonlySet<T> {
    if (a.size === 0 || b.size === 0) {
        return a;
    }

    const result = new Set<T>();
    for (const i of a) {
        if (!b.has(i)) {
            result.add(i);
        }
    }

    return result;
}

function setFilter<T>(a: Iterable<T>, fn: (item: T) => boolean): Set<T> {
    const result = new Set<T>();
    for (const i of a) {
        if (fn(i)) {
            result.add(i);
        }
    }
    return result;
}
