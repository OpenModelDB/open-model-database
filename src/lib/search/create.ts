import { deriveTags } from '../derive-tags';
import { Model, ModelId, TagId } from '../schema';
import { asArray } from '../util';
import { CorpusEntry, SearchIndex } from './search-index';

export function createModelSearchIndex(modelData: ReadonlyMap<ModelId, Model>) {
    return new SearchIndex(
        [...modelData].map(([id, model]): CorpusEntry<ModelId, TagId> => {
            return {
                id,
                tags: new Set(deriveTags(model)),
                texts: [
                    {
                        text: [id, model.name].filter(Boolean).join('\n').toLowerCase(),
                        weight: 8,
                    },
                    {
                        text: asArray(model.author).filter(Boolean).join('\n').toLowerCase(),
                        weight: 4,
                    },
                    {
                        text: [model.architecture, `${model.scale}x`, model.dataset]
                            .filter(Boolean)
                            .join('\n')
                            .toLowerCase(),
                        weight: 1,
                    },
                    { text: model.description.toLowerCase(), weight: 1 },
                ],
            };
        })
    );
}
