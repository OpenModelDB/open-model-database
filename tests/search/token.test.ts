import { tokenize } from '../../src/lib/search/token';

describe('tokenize', () => {
    const queries = [
        '',
        'foo',
        'fooBar',
        'FOO_BAR',
        'Foo Bar',
        'FooBar',
        '1x-AnimeUndeint-Compact',
        '1x-BleedOut-Compact',
        '1x-BroadcastToStudioLite',
        '1x-HurrDeblur-SuperUltraCompact',
        '1x-SpongeBC1-Lite',
        '2x-ATLA-KORRA',
        '2x-Bubble-AnimeScale-SwinIR-Small-v1',
        '2x-Byousoku-5-Centimeter',
        '2x-GT-evA',
        '2x-Loyaldk-LitePonyV2-0',
        '2x-SHARP-ANIME-V2',
        '2x-sudo-RealESRGAN-Dropout',
        '4x-BooruGan-650k',
        '2x-Waifaux-NL3-SuperLite',
        '1x-BaldrickVHSFixV0-2',
        '1x-BCGone-DetailedV2-40-60',
        '1x-BSChroma',
        '1x-GainRESV3-Natural',
        '1x-MangaJPEGLQ',
        '1x-ITF-SkinDiffDDS-v1',
        '1x-DEDXT',
    ];

    for (const query of queries) {
        test(query, () => {
            expect(tokenize(query)).toMatchSnapshot();
        });
    }
});
