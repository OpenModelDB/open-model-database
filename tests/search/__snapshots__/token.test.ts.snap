// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenize  1`] = `[]`;

exports[`tokenize 1x-AnimeUndeint-Compact 1`] = `
[
  "1x",
  "anime",
  "undeint",
  "compact",
]
`;

exports[`tokenize 1x-BCGone-DetailedV2-40-60 1`] = `
[
  "1x",
  "bc",
  "gone",
  "detailed",
  "v2",
  "40",
  "60",
]
`;

exports[`tokenize 1x-BSChroma 1`] = `
[
  "1x",
  "bs",
  "chroma",
]
`;

exports[`tokenize 1x-BaldrickVHSFixV0-2 1`] = `
[
  "1x",
  "baldrick",
  "vhs",
  "fix",
  "v0",
  "2",
]
`;

exports[`tokenize 1x-BleedOut-Compact 1`] = `
[
  "1x",
  "bleed",
  "out",
  "compact",
]
`;

exports[`tokenize 1x-BroadcastToStudioLite 1`] = `
[
  "1x",
  "broadcast",
  "to",
  "studio",
  "lite",
]
`;

exports[`tokenize 1x-DEDXT 1`] = `
[
  "1x",
  "de",
  "dxt",
]
`;

exports[`tokenize 1x-GainRESV3-Natural 1`] = `
[
  "1x",
  "gain",
  "res",
  "v3",
  "natural",
]
`;

exports[`tokenize 1x-HurrDeblur-SuperUltraCompact 1`] = `
[
  "1x",
  "hurr",
  "deblur",
  "super",
  "ultra",
  "compact",
]
`;

exports[`tokenize 1x-ITF-SkinDiffDDS-v1 1`] = `
[
  "1x",
  "itf",
  "skin",
  "diff",
  "dds",
  "v1",
]
`;

exports[`tokenize 1x-MangaJPEGLQ 1`] = `
[
  "1x",
  "manga",
  "jpeg",
  "lq",
]
`;

exports[`tokenize 1x-SpongeBC1-Lite 1`] = `
[
  "1x",
  "sponge",
  "bc",
  "1",
  "lite",
]
`;

exports[`tokenize 2x-ATLA-KORRA 1`] = `
[
  "2x",
  "atla",
  "korra",
]
`;

exports[`tokenize 2x-Bubble-AnimeScale-SwinIR-Small-v1 1`] = `
[
  "2x",
  "bubble",
  "anime",
  "scale",
  "swin",
  "ir",
  "small",
  "v1",
]
`;

exports[`tokenize 2x-Byousoku-5-Centimeter 1`] = `
[
  "2x",
  "byousoku",
  "5",
  "centimeter",
]
`;

exports[`tokenize 2x-GT-evA 1`] = `
[
  "2x",
  "gt",
  "eva",
]
`;

exports[`tokenize 2x-Loyaldk-LitePonyV2-0 1`] = `
[
  "2x",
  "loyaldk",
  "lite",
  "pony",
  "v2",
  "0",
]
`;

exports[`tokenize 2x-SHARP-ANIME-V2 1`] = `
[
  "2x",
  "sharp",
  "anime",
  "v2",
]
`;

exports[`tokenize 2x-Waifaux-NL3-SuperLite 1`] = `
[
  "2x",
  "waifaux",
  "nl",
  "3",
  "super",
  "lite",
]
`;

exports[`tokenize 2x-sudo-RealESRGAN-Dropout 1`] = `
[
  "2x",
  "sudo",
  "real",
  "esrgan",
  "dropout",
]
`;

exports[`tokenize 4x-BooruGan-650k 1`] = `
[
  "4x",
  "booru",
  "gan",
  "650",
  "k",
]
`;

exports[`tokenize FOO_BAR 1`] = `
[
  "foo",
  "bar",
]
`;

exports[`tokenize Foo Bar 1`] = `
[
  "foo",
  "bar",
]
`;

exports[`tokenize FooBar 1`] = `
[
  "foo",
  "bar",
]
`;

exports[`tokenize foo 1`] = `
[
  "foo",
]
`;

exports[`tokenize fooBar 1`] = `
[
  "foo",
  "bar",
]
`;
